cmd/notation/internal/plugin/plugin.go

```diff
@@ -0,0 +1,179 @@
+)
+
+const (
+	// MediaTypeZip means plugin file is zip
+	MediaTypeZip = "application/zip"
+
+	// MediaTypeGzip means plugin file is gzip
+	MediaTypeGzip = "application/x-gzip"
+)
+
+// GetPluginMetadataIfExist returns plugin's metadata if it exists in Notation
+func GetPluginMetadataIfExist(ctx context.Context, pluginName string) (*proto.GetMetadataResponse, error) {
+	mgr := plugin.NewCLIManager(dir.PluginFS())
+	plugin, err := mgr.Get(ctx, pluginName)
+	if err != nil {
+		return nil, err
+	}
+	return plugin.GetMetadata(ctx, &proto.GetMetadataRequest{})
+}
+
+// GetPluginMetadata returns plugin's metadata given plugin path
+func GetPluginMetadata(ctx context.Context, pluginName, path string) (*proto.GetMetadataResponse, error) {
+	plugin, err := plugin.NewCLIPlugin(ctx, pluginName, path)
+	if err != nil {
+		return nil, err
+	}
+	return plugin.GetMetadata(ctx, &proto.GetMetadataRequest{})
+}
+
+// ComparePluginVersion validates and compares two plugin semantic versions
+func ComparePluginVersion(v, w string) (int, error) {
+	// semantic version strings must begin with a leading "v"
+	if !strings.HasPrefix(v, "v") {
+		v = "v" + v
+	}
+	if !semver.IsValid(v) {
+		return 0, fmt.Errorf("%s is not a valid semantic version", v)
+	}
+	if !strings.HasPrefix(w, "v") {
+		w = "v" + w
+	}
+	if !semver.IsValid(w) {
+		return 0, fmt.Errorf("%s is not a valid semantic version", w)
+	}
+	return semver.Compare(v, w), nil
+}
+
+// ValidateCheckSum returns nil if SHA256 of file at path equals to checkSum.
+func ValidateCheckSum(path string, checkSum string) error {
+	rc, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer rc.Close()
+	dgst, err := digest.FromReader(rc)
+	if err != nil {
+		return err
+	}
+	enc := dgst.Encoded()
+	if enc != strings.ToLower(checkSum) {
+		return fmt.Errorf("plugin checksum does not match user input. Expecting %s", checkSum)
+	}
+	return nil
+}
+
+// DownloadPluginFromURL downloads plugin file from url to a tmp directory
+func DownloadPluginFromURL(ctx context.Context, url string, tmpFile *os.File) error {
+	// Get the data
+	client := getClient(ctx)
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return err
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	// Check server response
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("bad status: %s", resp.Status)
+	}
+	// Write the body to file
+	lr := io.LimitReader(resp.Body, maxPluginSourceBytes)
+	_, err = io.Copy(tmpFile, lr)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// getClient returns an *auth.Client
+func getClient(ctx context.Context) *auth.Client {
+	client := &auth.Client{
+		Cache:    auth.NewCache(),
+		ClientID: "notation",
+	}
+	client.SetUserAgent("notation/" + version.GetVersion())
+	setHttpDebugLog(ctx, client)
+	return client
+}
+
+// setHttpDebugLog sets http debug logger
+func setHttpDebugLog(ctx context.Context, authClient *auth.Client) {
```

Comment:
This is a suggestion from @shizhMSFT. The main purpose here is to reuse the code we already have. I will need to refactor `getAuthClient` under `main` package so that it can be reused in other packages. I also moved `setHttpDebugLog` into the `trace` package since it's not really registry specific.